kernel/src/hil/asymmetric_encryption.rs

```diff
@@ -0,0 +1,132 @@
+//! Interface for asymmetric-cipher encryption
+
+use crate::hil::entropy;
+use crate::ErrorCode;
+
+/// Upcall from the `PubPrivKey` trait.
+pub trait PubPrivKeyClient<'a> {
+    /// The `generate()` command has been completed.
+    fn generation_complete(&'a self, result: Result<(), ErrorCode>);
+}
+
+/// An internal representation of a asymetric Public and/or Private key.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubPrivKey<'a> {
+    /// Set the client. This client will be called when the `gnerate()`
+    /// function is complete. If using an existing key this doesn't need to be
+    /// used.
+    fn set_client(&'a self, client: &'a dyn PubPrivKeyClient<'a>);
+
+    /// This generates a new private/public key pair of `length` bits.
+    /// This will call the `generation_complete()` on completion. They keys
+    /// cannot be used and will return `None` until the upcall has been called.
+    ///
+    /// The keys generated by `generate()` will depend on the implementation.
+    fn gnerate(&'a self, trng: &'a dyn entropy::Entropy32, length: usize) -> Result<(), ErrorCode>;
+
+    /// Import an existing public key. This will override any existing keys.
+    fn import_public_key(&mut self, public_key: &[u8]) -> Result<(), ErrorCode>;
```

Comment:
Should this be typed, rather than a slice? I.e., do you expect to have variable-length keys?

Reply:
I expect variable length keys between implementations, but not in the same implementation. I should be able to change this to use const generics.

Reply:
Actually no, I do expect variable length keys. For example RSA could have 2048 or 4096 bit keys

Reply:
I'd argue those are not variable-length keys; those are two specific lengths which can be expected. I.e., you don't have 1054-bit keys, or 4000-bit keys. Should there be stronger typing here?

Reply:
What are the valid ErrorCodes and what do they mean?